# Task H - Lift Analysis
<p>Author: Samuel D'Avila</p>

In [1]:
# Importing libraries

import numpy as np
import pandas as pd
import ReferenceArchitecture as ra

In [2]:
# Import CSV scores
# can do check if for file if I have more time
df_sim_scores = pd.read_csv('similarity_scores.csv')

In [3]:
df_sim_scores

Unnamed: 0                 product_name  \
0              1  Kentucky Brunch Brand Stout   
1              2  Kentucky Brunch Brand Stout   
2              3  Kentucky Brunch Brand Stout   
3              4  Kentucky Brunch Brand Stout   
4              5  Kentucky Brunch Brand Stout   
...          ...                          ...   
2508        2509                 Coolship Red   
2509        2510                 Coolship Red   
2510        2511                 Coolship Red   
2511        2512                 Coolship Red   
2512        2513                 Coolship Red   

                                         product_review  similarity_score  
0     I didnt think i was going to give it a perfect...          0.035921  
1     So I just read a review that called the legend...          0.019702  
2     2021 vintage bottle 79\n\nHoly Fucking Shit KB...          0.110674  
3     Celebrating my buddy Rug with his 1000th beer ...          0.071157  
4     Thick and syrupy pour mocha head Aroma is booz...          0.063246  
...                                                 ...               ...  
2508  Very interesting and complex ale First this is...          0.000000  
2509  Appearance  The beer was pinkish red is color ...          0.000000  
2510  Look beautiful opaque garnet red Soft white he...          0.000000  
2511  Caged and corked bottle dated dec32020 and ope...          0.000000  
2512  I want more funk The raspberry character is bi...          0.000000  

[2513 rows x 4 columns]

In [4]:
df_sim_scores['name_review'] = df_sim_scores['product_name'] + " " + df_sim_scores['product_review']
df_sim_scores

Unnamed: 0                 product_name  \
0              1  Kentucky Brunch Brand Stout   
1              2  Kentucky Brunch Brand Stout   
2              3  Kentucky Brunch Brand Stout   
3              4  Kentucky Brunch Brand Stout   
4              5  Kentucky Brunch Brand Stout   
...          ...                          ...   
2508        2509                 Coolship Red   
2509        2510                 Coolship Red   
2510        2511                 Coolship Red   
2511        2512                 Coolship Red   
2512        2513                 Coolship Red   

                                         product_review  similarity_score  \
0     I didnt think i was going to give it a perfect...          0.035921   
1     So I just read a review that called the legend...          0.019702   
2     2021 vintage bottle 79\n\nHoly Fucking Shit KB...          0.110674   
3     Celebrating my buddy Rug with his 1000th beer ...          0.071157   
4     Thick and syrupy pour mocha head Aroma is booz...          0.063246   
...                                                 ...               ...   
2508  Very interesting and complex ale First this is...          0.000000   
2509  Appearance  The beer was pinkish red is color ...          0.000000   
2510  Look beautiful opaque garnet red Soft white he...          0.000000   
2511  Caged and corked bottle dated dec32020 and ope...          0.000000   
2512  I want more funk The raspberry character is bi...          0.000000   

                                            name_review  
0     Kentucky Brunch Brand Stout I didnt think i wa...  
1     Kentucky Brunch Brand Stout So I just read a r...  
2     Kentucky Brunch Brand Stout 2021 vintage bottl...  
3     Kentucky Brunch Brand Stout Celebrating my bud...  
4     Kentucky Brunch Brand Stout Thick and syrupy p...  
...                                                 ...  
2508  Coolship Red Very interesting and complex ale ...  
2509  Coolship Red Appearance  The beer was pinkish ...  
2510  Coolship Red Look beautiful opaque garnet red ...  
2511  Coolship Red Caged and corked bottle dated dec...  
2512  Coolship Red I want more funk The raspberry ch...  

[2513 rows x 5 columns]

In [5]:
beer_attributes = ['chocolate', 'dark', 'sweet', 'bourbon']

all_beer_names = [i for i in np.unique(df_sim_scores['product_name'])]
rand_beers = [all_beer_names[np.random.randint(0,len(all_beer_names))] for i in range(0, 9, 1)]

beer_attributes_names = [i for i in beer_attributes]
for i in rand_beers:
    beer_attributes_names.append(i)

col = 'name_review'
lift_table, lift_dict_values = ra.lift_ratio(index=beer_attributes_names, data=df_sim_scores, capture_column=col)

In [6]:
lift_table

In [7]:
lift_dict_values

chocolate    dark  sweet bourbon  \
chocolate                                       0  15.206  5.805  18.721   
dark                                       15.206       0  5.563  12.391   
sweet                                       5.805   5.563      0   4.618   
bourbon                                    18.721  12.391  4.618       0   
Atrial Rubicite                                 0       0      0       0   
The Adjunct Trail - Bourbon Barrel-Aged         0       0      0       0   
Double Dry Hopped Double Citra Daydream         0       0      0       0   
Clare's Thirsty Ale                         3.378   2.336  0.935   3.051   
Coconut B-Bomb                              2.167    1.36  0.415   3.051   
Affogato - Bourbon Barrel-Aged                  0   1.357  0.934    1.77   
Juice Machine                                   0   0.837  1.878       0   
Henna: Barrel-Aged Double Rainbow           3.792   0.905      0    2.84   
JJJuliusss!                                     0    1.36  5.643       0   

                                        Atrial Rubicite  \
chocolate                                             0   
dark                                                  0   
sweet                                                 0   
bourbon                                               0   
Atrial Rubicite                                       0   
The Adjunct Trail - Bourbon Barrel-Aged               0   
Double Dry Hopped Double Citra Daydream               0   
Clare's Thirsty Ale                                   0   
Coconut B-Bomb                                        0   
Affogato - Bourbon Barrel-Aged                        0   
Juice Machine                                         0   
Henna: Barrel-Aged Double Rainbow                     0   
JJJuliusss!                                           0   

                                        The Adjunct Trail - Bourbon Barrel-Aged  \
chocolate                                                                     0   
dark                                                                          0   
sweet                                                                         0   
bourbon                                                                       0   
Atrial Rubicite                                                               0   
The Adjunct Trail - Bourbon Barrel-Aged                                       0   
Double Dry Hopped Double Citra Daydream                                       0   
Clare's Thirsty Ale                                                           0   
Coconut B-Bomb                                                                0   
Affogato - Bourbon Barrel-Aged                                                0   
Juice Machine                                                                 0   
Henna: Barrel-Aged Double Rainbow                                             0   
JJJuliusss!                                                                   0   

                                        Double Dry Hopped Double Citra Daydream  \
chocolate                                                                     0   
dark                                                                          0   
sweet                                                                         0   
bourbon                                                                       0   
Atrial Rubicite                                                               0   
The Adjunct Trail - Bourbon Barrel-Aged                                       0   
Double Dry Hopped Double Citra Daydream                                       0   
Clare's Thirsty Ale                                                           0   
Coconut B-Bomb                                                                0   
Affogato - Bourbon Barrel-Aged                                                0   
Juice Machine                                                           

In [8]:
rando_beer = lift_dict_values.columns[np.random.randint(0, len(lift_dict_values))]
rando_beer

'Affogato - Bourbon Barrel-Aged'

In [29]:
lift_dict_values.columns

Index(['chocolate', 'dark', 'sweet', 'bourbon', 'Atrial Rubicite',
       'The Adjunct Trail - Bourbon Barrel-Aged',
       'Double Dry Hopped Double Citra Daydream', 'Clare's Thirsty Ale',
       'Coconut B-Bomb', 'Affogato - Bourbon Barrel-Aged', 'Juice Machine',
       'Henna: Barrel-Aged Double Rainbow', 'JJJuliusss!'],
      dtype='object')

In [27]:
diff = []
idx = []
for i in lift_dict_values.columns[4:]:
    if i == rando_beer:
        continue
    series = abs(lift_dict_values[rando_beer][:4]) - abs(lift_dict_values[i][:4])
    val = sum(series.values)
    diff.append(val)
    idx.append(i)


In [28]:
most_sim = pd.Series(diff, index=idx)
most_sim

Atrial Rubicite                            4.061
The Adjunct Trail - Bourbon Barrel-Aged    4.061
Double Dry Hopped Double Citra Daydream    4.061
Clare's Thirsty Ale                       -5.639
Coconut B-Bomb                            -2.932
Juice Machine                              1.346
Henna: Barrel-Aged Double Rainbow         -3.476
JJJuliusss!                               -2.942
dtype: float64

For our code, we randomly select 10 beers and then randomly select 1 beer from those ten. This random beer is compared to others through a taking the lift values across the 4 attributes (in our case: "chocolate," "dark," "sweet," "bourbon") and summing the whole series of values for each beer (we now have 10 beers with a 1 value that represents the sum of their lift values on their attributes). We then take the difference between the combined lift for the selected beer and the remaining beers. The closer the remaining values are to zero, the more similar that one of the nine beers is the selected beer. During the code's testing, we had the following 10 beers:
    'Atrial Rubicite',
    'The Adjunct Trail - Bourbon Barrel-Aged',
    'Double Dry Hopped Double Citra Daydream',
    'Clare's Thirsty Ale',
    'Coconut B-Bomb',
    'Affogato - Bourbon Barrel-Aged', 
    'Juice Machine',
    'Henna: Barrel-Aged Double Rainbow', 
    'JJJuliusss!'. 
The randomly selected beer was the 'Affogato - Bourbon Barrel-Aged' and the following lift differences were calculated for each beer:
Atrial Rubicite                            4.061
The Adjunct Trail - Bourbon Barrel-Aged    4.061
Double Dry Hopped Double Citra Daydream    4.061
Clare's Thirsty Ale                       -5.639
Coconut B-Bomb                            -2.932
Juice Machine                              1.346
Henna: Barrel-Aged Double Rainbow         -3.476
JJJuliusss!                               -2.942

From these values we can say that the 'Juice Machine' is most similar to 'Affogato - Bourbon Barrel-Aged' beer.